File: configuration-manager-go-client/user_service.go
package configmanager

import (
	"context"
	"fmt"
)

// GetAll retrieves all users (admin only)
func (u *UserService) GetAll(ctx context.Context) (*UsersListResponse, error) {
	resp, err := u.client.Get(ctx, "/users")
	if err != nil {
		return nil, fmt.Errorf("get users request failed: %w", err)
	}
	defer u.client.CloseResponse(resp)

	var response UsersListResponse
	if err := u.client.DecodeJSON(resp, &response); err != nil {
		return nil, fmt.Errorf("failed to decode users response: %w", err)
	}

	return &response, nil
}

// Get retrieves a specific user by ID
func (u *UserService) Get(ctx context.Context, id string) (*UserResponse, error) {
	if id == "" {
		return nil, NewValidationError("user ID cannot be empty", 0)
	}

	path := fmt.Sprintf("/users/%s", id)
	resp, err := u.client.Get(ctx, path)
	if err != nil {
		return nil, fmt.Errorf("get user request failed: %w", err)
	}
	defer u.client.CloseResponse(resp)

	var response UserResponse
	if err := u.client.DecodeJSON(resp, &response); err != nil {
		return nil, fmt.Errorf("failed to decode user response: %w", err)
	}

	return &response, nil
}

// UpdateRole updates user role (admin only)
func (u *UserService) UpdateRole(ctx context.Context, id, role string) (*UserRoleUpdateResponse, error) {
	if id == "" {
		return nil, NewValidationError("user ID cannot be empty", 0)
	}

	if role == "" {
		return nil, NewValidationError("role cannot be empty", 0)
	}

	if role != "USER" && role != "ADMIN" {
		return nil, NewValidationError("role must be either 'USER' or 'ADMIN'", 0)
	}

	request := &UpdateUserRoleRequest{Role: role}

	path := fmt.Sprintf("/users/%s/role", id)
	resp, err := u.client.Put(ctx, path, request)
	if err != nil {
		return nil, fmt.Errorf("update user role request failed: %w", err)
	}
	defer u.client.CloseResponse(resp)

	var response UserRoleUpdateResponse
	if err := u.client.DecodeJSON(resp, &response); err != nil {
		return nil, fmt.Errorf("failed to decode update user role response: %w", err)
	}

	return &response, nil
}

// Delete deletes a user (admin only)
func (u *UserService) Delete(ctx context.Context, id string) (*UserDeletionResponse, error) {
	if id == "" {
		return nil, NewValidationError("user ID cannot be empty", 0)
	}

	path := fmt.Sprintf("/users/%s", id)
	resp, err := u.client.Delete(ctx, path)
	if err != nil {
		return nil, fmt.Errorf("delete user request failed: %w", err)
	}
	defer u.client.CloseResponse(resp)

	var response UserDeletionResponse
	if err := u.client.DecodeJSON(resp, &response); err != nil {
		return nil, fmt.Errorf("failed to decode delete user response: %w", err)
	}

	return &response, nil
}

// GetConfigurations gets configurations created by a user
func (u *UserService) GetConfigurations(ctx context.Context, id string) (*UserConfigurationsResponse, error) {
	if id == "" {
		return nil, NewValidationError("user ID cannot be empty", 0)
	}

	path := fmt.Sprintf("/users/%s/configurations", id)
	resp, err := u.client.Get(ctx, path)
	if err != nil {
		return nil, fmt.Errorf("get user configurations request failed: %w", err)
	}
	defer u.client.CloseResponse(resp)

	var response UserConfigurationsResponse
	if err := u.client.DecodeJSON(resp, &response); err != nil {
		return nil, fmt.Errorf("failed to decode user configurations response: %w", err)
	}

	return &response, nil
}

File: configuration-manager-python-client/configuration_manager_client/services/user_service.py
"""User management service for Configuration Manager API."""

from ..http_client import HTTPClient
from ..models.users import (
    UsersListResponse,
    UserResponse,
    UpdateUserRoleRequest,
    UserRoleUpdateResponse,
    UserDeletionResponse,
    UserConfigurationsResponse,
)
from ..exceptions import ValidationError
from ..utils import from_dict


class UserService:
    """User management service client."""

    def __init__(self, http_client: HTTPClient):
        """Initialize user service.
        
        Args:
            http_client: HTTP client instance
        """
        self.client = http_client

    def get_all(self) -> UsersListResponse:
        """Get all users (admin only).
        
        Returns:
            UsersListResponse containing list of users
        """
        response = self.client.get("/users")
        return from_dict(UsersListResponse, response.json())

    def get(self, user_id: str) -> UserResponse:
        """Get a specific user by ID.
        
        Args:
            user_id: User ID
            
        Returns:
            UserResponse containing user information
            
        Raises:
            ValidationError: If user_id is empty
        """
        if not user_id:
            raise ValidationError("User ID cannot be empty")

        response = self.client.get(f"/users/{user_id}")
        return from_dict(UserResponse, response.json())

    def update_role(self, user_id: str, role: str) -> UserRoleUpdateResponse:
        """Update user role (admin only).
        
        Args:
            user_id: User ID
            role: New role ("USER" or "ADMIN")
            
        Returns:
            UserRoleUpdateResponse containing updated user
            
        Raises:
            ValidationError: If parameters are invalid
        """
        if not user_id:
            raise ValidationError("User ID cannot be empty")

        if not role:
            raise ValidationError("Role cannot be empty")

        if role not in ("USER", "ADMIN"):
            raise ValidationError("Role must be either 'USER' or 'ADMIN'")

        data = {"role": role}
        response = self.client.put(f"/users/{user_id}/role", json_data=data)
        return from_dict(UserRoleUpdateResponse, response.json())

    def delete(self, user_id: str) -> UserDeletionResponse:
        """Delete a user (admin only).
        
        Args:
            user_id: User ID
            
        Returns:
            UserDeletionResponse containing deletion confirmation
            
        Raises:
            ValidationError: If user_id is empty
        """
        if not user_id:
            raise ValidationError("User ID cannot be empty")

        response = self.client.delete(f"/users/{user_id}")
        return from_dict(UserDeletionResponse, response.json())

    def get_configurations(self, user_id: str) -> UserConfigurationsResponse:
        """Get configurations created by a user.
        
        Args:
            user_id: User ID
            
        Returns:
            UserConfigurationsResponse containing user's configurations
            
        Raises:
            ValidationError: If user_id is empty
        """
        if not user_id:
            raise ValidationError("User ID cannot be empty")

        response = self.client.get(f"/users/{user_id}/configurations")
        return from_dict(UserConfigurationsResponse, response.json())
